src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs

```diff
@@ -126,69 +117,93 @@ public WebSocketsTransport(WebSocketOptions options, IChannelConnection<Message>
             }
         }
 
-        private Task HandleFrame(WebSocketFrame frame)
+        private async Task<WebSocketReceiveResult> StartReceiving(WebSocket socket)
         {
-            // Is this a frame we care about?
-            if (!frame.Opcode.IsMessage())
+            // REVIEW: This code was copied from the client, it's highly unoptimized at the moment (especially 
+            // for server logic)
+            var incomingMessage = new List<ArraySegment<byte>>();
+            while (true)
             {
-                return Task.CompletedTask;
-            }
+                const int bufferSize = 4096;
+                var totalBytes = 0;
+                WebSocketReceiveResult receiveResult;
+                do
+                {
+                    var buffer = new ArraySegment<byte>(new byte[bufferSize]);
 
-            LogFrame("Receiving", frame);
+                    // Exceptions are handled above where the send and receive tasks are being run.
+                    receiveResult = await socket.ReceiveAsync(buffer, CancellationToken.None);
+                    if (receiveResult.MessageType == WebSocketMessageType.Close)
+                    {
+                        return receiveResult;
+                    }
 
-            // Determine the effective opcode based on the continuation.
-            var effectiveOpcode = frame.Opcode;
-            if (frame.Opcode == WebSocketOpcode.Continuation)
-            {
-                effectiveOpcode = _lastOpcode;
-            }
-            else
-            {
-                _lastOpcode = frame.Opcode;
-            }
+                    _logger.LogDebug("Message received. Type: {messageType}, size: {size}, EndOfMessage: {endOfMessage}",
+                        receiveResult.MessageType, receiveResult.Count, receiveResult.EndOfMessage);
 
-            // Create a Message for the frame
-            // This has to copy the buffer :(.
-            var message = new Message(frame.Payload.ToArray(), effectiveOpcode == WebSocketOpcode.Binary ? MessageType.Binary : MessageType.Text, frame.EndOfMessage);
+                    var truncBuffer = new ArraySegment<byte>(buffer.Array, 0, receiveResult.Count);
+                    incomingMessage.Add(truncBuffer);
+                    totalBytes += receiveResult.Count;
+                } while (!receiveResult.EndOfMessage);
 
-            // Write the message to the channel
-            return _application.Output.WriteAsync(message);
-        }
+                // Making sure the message type is either text or binary
+                Debug.Assert((receiveResult.MessageType == WebSocketMessageType.Binary || receiveResult.MessageType == WebSocketMessageType.Text), "Unexpected message type");
```

Comment:
This isn't going to work for multi-part messages. Only the first message will have a `Binary` or `Text` message type, the remaining messages are of type `Continuation`. You need to capture the type of the first frame of the message for use here.